File: Contents/code/haskel/chapter5/data_transformation.hs
module DataTransformation where

-- 1. Mapping (Data Transformation)
-- Applying a function to every item in a list.

numbers :: [Int]
numbers = [1, 2, 3, 4, 5]

square :: Int -> Int
square x = x * x

-- Using built-in map
-- Type signature of map: (a -> b) -> [a] -> [b]
squaredNumbersMap :: [Int]
squaredNumbersMap = map square numbers
-- Expected: [1, 4, 9, 16, 25]

-- Pythonic alternative equivalent: List Comprehension
squaredNumbersComp :: [Int]
squaredNumbersComp = [square x | x <- numbers]


-- 2. Filtering
-- Keeping items that satisfy a predicate.

isEven :: Int -> Bool
isEven x = x `mod` 2 == 0

-- Using built-in filter
-- Type signature of filter: (a -> Bool) -> [a] -> [a]
evenNumbersFilter :: [Int]
evenNumbersFilter = filter isEven numbers
-- Expected: [2, 4]

-- List Comprehension with condition
evenNumbersComp :: [Int]
evenNumbersComp = [x | x <- numbers, isEven x]


-- 3. Reducing (Folding)
-- Accumulating items down to a single value.

add :: Int -> Int -> Int
add x y = x + y

-- Using foldl (left fold)
-- Type signature of foldl: (b -> a -> b) -> b -> [a] -> b
-- It takes an accumulator function, an initial value, and a list.
sumFoldl :: Int
sumFoldl = foldl add 0 numbers
-- Expected: 15

-- Using foldr (right fold)
-- Type signature of foldr: (a -> b -> b) -> b -> [a] -> b
sumFoldr :: Int
sumFoldr = foldr add 0 numbers
-- Expected: 15


-- 4. Defining our own Map and Filter with recursion

myMap :: (a -> b) -> [a] -> [b]
myMap _ [] = []
myMap f (x:xs) = f x : myMap f xs

myFilter :: (a -> Bool) -> [a] -> [a]
myFilter _ [] = []
myFilter p (x:xs)
    | p x       = x : myFilter p xs
    | otherwise = myFilter p xs


-- 5. Combining Map and Filter
-- Square only the even numbers

-- Functional composition style (.)
-- Note: Function composition is read right-to-left
squaredEvensFunc :: [Int]
squaredEvensFunc = map square (filter isEven numbers)

-- Using the composition operator
squaredEvensCompose :: [Int]
squaredEvensCompose = (map square . filter isEven) numbers

-- List Comprehension style
squaredEvensComp :: [Int]
squaredEvensComp = [square x | x <- numbers, isEven x]

main :: IO ()
main = do
    putStrLn $ "Original: " ++ show numbers
    putStrLn $ "Map (square): " ++ show squaredNumbersMap
    putStrLn $ "Filter (even): " ++ show evenNumbersFilter
    putStrLn $ "Fold (sum): " ++ show sumFoldl
    putStrLn $ "Combined (square evens): " ++ show squaredEvensFunc

File: Contents/code/python/chapter5/data_transformation.py
from typing import List, Callable, TypeVar, Iterable
from functools import reduce

A = TypeVar('A')
B = TypeVar('B')

# 1. Mapping (Data Transformation)
# Applying a function to every item in an iterable.

numbers: List[int] = [1, 2, 3, 4, 5]

# Using built-in map
def square(x: int) -> int:
    return x * x

squared_numbers_map: List[int] = list(map(square, numbers))
# Expected: [1, 4, 9, 16, 25]

# Pythonic alternative: List Comprehension
squared_numbers_comp: List[int] = [square(x) for x in numbers]


# 2. Filtering
# Keeping items that satisfy a predicate.

def is_even(x: int) -> bool:
    return x % 2 == 0

# Using built-in filter
even_numbers_filter: List[int] = list(filter(is_even, numbers))
# Expected: [2, 4]

# Pythonic alternative: List Comprehension with condition
even_numbers_comp: List[int] = [x for x in numbers if is_even(x)]


# 3. Reducing (Folding)
# Accumulating items down to a single value.

def add(x: int, y: int) -> int:
    return x + y

# Using functools.reduce
sum_reduce: int = reduce(add, numbers)
# Expected: 15 (1+2+3+4+5)

# With an initial value
sum_reduce_initial: int = reduce(add, numbers, 10)
# Expected: 25 (10+1+2+3+4+5)

# Pythonic alternative: Built-in function
sum_builtin: int = sum(numbers)


# 4. Defining our own Map and Filter to see how they work

def my_map(func: Callable[[A], B], iterable: Iterable[A]) -> List[B]:
    result : List[B] = []
    for item in iterable:
        result.append(func(item))
    return result

def my_filter(predicate: Callable[[A], bool], iterable: Iterable[A]) -> List[A]:
    result : List[A] = []
    for item in iterable:
        if predicate(item):
            result.append(item)
    return result

# 5. Combining Map and Filter
# Square only the even numbers

# Functional style
squared_evens_func: List[int] = list(map(square, filter(is_even, numbers)))

# List Comprehension style
squared_evens_comp: List[int] = [square(x) for x in numbers if is_even(x)]

if __name__ == "__main__":
    print(f"Original: {numbers}")
    print(f"Map (square): {squared_numbers_map}")
    print(f"Filter (even): {even_numbers_filter}")
    print(f"Reduce (sum): {sum_reduce}")
    print(f"Combined (square evens): {squared_evens_func}")
